### PR TITLE
[Frontend] Bad words sampling parameter

### LINKED ISSUE #986: It seems that SamplingParams doesnt support the bad_words_ids parameter when generating
`bad_words_ids` described [here](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/configuration_utils.py#L145C11-L145C11) is useful for production applications.
However It seems that vlllm doesnt support the bad_words_ids parameter when generating.
Is there a plan to support it?
```
  bad_words_ids(`List[List[int]]`, *optional*):
  List of list of token ids that are not allowed to be generated. Check
  [`~generation.NoBadWordsLogitsProcessor`] for further documentation and examples.
```

### PR DESCRIPTION
**Pull request by Alvant**
For comments and discussion, see this PR: https://github.com/vllm-project/vllm/pull/5986.

FIX #986

### CODE REVIEW

File: vllm/logits_process.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import torch
+
+LogitsProcessor = Union[Callable[[List[int], torch.Tensor], torch.Tensor],
+                        Callable[[List[int], List[int], torch.Tensor],
+                                 torch.Tensor]]
+"""LogitsProcessor is a function that takes a list
+of previously generated tokens, the logits tensor
+for the next token and, optionally, prompt tokens as a
+first argument, and returns a modified tensor of logits
+to sample from."""
```
Comment (DarkLight1337) [inline vllm/logits_process.py:12]: Let's import this from `vllm.sampling_params` instead of copying the definition.

File: vllm/engine/llm_engine.py [hunk 2]
```diff
@@ -723,6 +725,44 @@ def _add_processed_request(
     def stop_remote_worker_execution_loop(self) -> None:
         self.model_executor.stop_remote_worker_execution_loop()
 
+    def process_model_params(
```
Comment (DarkLight1337) [inline vllm/engine/llm_engine.py:728]: I suggest adding this to `_build_logits_processors` instead of creating a new method. There is also no need to create an async version.

**Comment by DarkLight1337** [comment]
Thanks for adding this! Some comments above.

Also, don't worry about DCO - we can manually pass it if needed.

File: vllm/engine/llm_engine.py (hunk 2 above)
Comment (Alvant) [reply vllm/engine/llm_engine.py]: Ok! I didn't notice `_build_logits_processors`. Indeed, it seems the right place for the "bad words". Moved logic here. Also removed the async version. I guess that really was an overkill)

File: vllm/engine/llm_engine.py [hunk 3]
```diff
@@ -1994,6 +1998,32 @@ def _build_logits_processors(
...
+                    # If no space at the beginning
+                    # or if prefix space produces a new word token
+                    if (not add_prefix_space) or (
+                            add_prefix_space
+                            and prompt_token_ids[0] != bad_words_ids[-1][0] and
+                            len(prompt_token_ids) == len(bad_words_ids[-1])):
+                        bad_words_ids.append(prompt_token_ids)
+
+            no_bad_words_processor = NoBadWordsLogitsProcessor(
+                bad_words_ids=bad_words_ids)
```
Comment (DarkLight1337) [inline vllm/engine/llm_engine.py:2024]: Perhaps we can factor out this code? Similar to `get_logits_processors`

File: vllm/logits_process.py (hunk 1 above)
Comment (Alvant) [reply vllm/logits_process.py:12]: It is not copied, it is just moved from `vllm.sampling_params` to this file. Is it Ok?

It seemed more logical to me to place this `LogitsProcessor` in a file near the newly introduced `NoBadWordsLogitsProcessor` (all these "base" logits processors in one place).

File: vllm/engine/llm_engine.py [hunk 4]
```diff
@@ -707,8 +711,7 @@ def _add_processed_request(
                 encoder_seq=encoder_seq,
                 priority=priority)
         else:
-            raise ValueError(
-                "Either SamplingParams or PoolingParams must be provided.")
```
Comment (DarkLight1337) [inline vllm/engine/llm_engine.py:711]: No need to update this code now that this is the only place where `INCORRECT_PARAMS_TYPE_MSG` is used.

File: vllm/logits_process.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/logits_process.py:12]: I see what you mean now. That's fine then.

File: vllm/engine/llm_engine.py (hunk 3 above)
Comment (Alvant) [reply vllm/engine/llm_engine.py]: Ok, I think I get the idea... Tried to unify the code.

File: vllm/logits_process.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+
+def get_bad_words_logits_processors(
+        bad_words: List[str],
+        tokenizer: AnyTokenizer) -> List[LogitsProcessor]:
```
Comment (Alvant) [inline vllm/logits_process.py:19]: I wanted to specify the type of the tokenizer here. But found out (during PR auto code checks) that... `MistralTokenizer` has a different `encode`'s signature, so I had to add a tokenizer class check below to handle Mistral's case differently...

File: vllm/engine/llm_engine.py [hunk 6]
```diff
@@ -1994,6 +1997,31 @@ def _build_logits_processors(
...
+                    prompt_token_ids = tokenizer.encode(
+                        text=inputs["prompt"], add_special_tokens=False)
+
+                    # If no space at the beginning
+                    # or if prefix space produces a new word token
+                    if (not add_prefix_space) or (
+                            add_prefix_space
+                            and prompt_token_ids[0] != bad_words_ids[-1][0] and
+                            len(prompt_token_ids) == len(bad_words_ids[-1])):
+                        bad_words_ids.append(prompt_token_ids)
```
[the remark this answers is not in the document]
Comment (Alvant) [reply vllm/engine/llm_engine.py]: Moved all logits creation stuff to `get_logits_processors  :+1: 

Rename the method to `get_bad_words_logits_processors`?.. :thinking: Well... I do not believe that there ever will be other logits processors related to bad words) So plural case is a bit overkill in this context, in my opinion. But, well, let it be so, with the name like that to make things clearer...

File: vllm/logits_process.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    bad_words_ids: List[List[int]] = list()
+
+    for bad_word in bad_words:
+        # To prohibit words both at the beginning
+        # and in the middle of text
+        # (related to add_prefix_space tokenizer parameter)
+        for add_prefix_space in [False, True]:
+            prefix = " " if add_prefix_space else ""
+            inputs = {"prompt": prefix + bad_word.lstrip()}
```
Comment (DarkLight1337) [inline vllm/logits_process.py:28]: Is there a particular reason why we have to make inputs a dictionary here?

File: vllm/logits_process.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        if self.word_bias is None:
+            self._init_word_bias(logits=logits)
+
+        last_token_bias = torch.zeros_like(logits)
+
+        for bad_word_ids in self.bad_words_ids:
+            if len(bad_word_ids) == 1:  # 1-token words already processed
+                continue
+
+            if len(bad_word_ids) > len(past_tokens_ids) + 1:
```
Comment (Alvant) [inline vllm/logits_process.py:70]: Since the original PR has faded into history, I want to bring to the surface one of the notes which I believe might be important (at least to the point where it's worth highlighting it again)

Comment (Alvant) [inline vllm/logits_process.py:70]: This differs from the original inequality here: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py#L1131
```python
if len(sequence_ids) > input_ids.shape[1]:  # the sequence is longer than the context, ignore
    continue
```

I may be wrong, but it seemed we should allow for one token to be generated (so + 1).

Comment (DarkLight1337) [reply vllm/logits_process.py:70]: According to your interpretation, we should also change `prefix_length` to `len(bad_word_ids)`. See if your tests still pass with this change.

File: vllm/logits_process.py (hunk 7 above)
Comment (Alvant) [reply vllm/logits_process.py]: I look at the code now and can't remember any) Probably, I just copy-pasted the lines related to the tokenizer's encoding from some other place in vLLM... So removed the dictionary, keep just a string `prompt`.

File: vllm/logits_process.py (hunk 8 above)
Comment (Alvant) [reply vllm/logits_process.py:70]: One passes, one fails:



Well, I think, I better try come up with some illustration of the idea (to make the picture clearer for everybody, including me :slightly_smiling_face:)

Suppose, `past_tokens_ids = ["bye"]` (well, that's not ids, but words — let's go with it for simplicity).
And `bad_word_ids = ["bye", "world"]`.
So, we do not want "bye world" to be generated, but we already have "bye" generated.

What is happening now:
```python
past_tokens_ids = ["bye"]
bad_word_ids = ["bye", "world"]

if len(bad_word_ids) > len(past_tokens_ids) + 1:  # 2 > 1 + 1 -> False
                                                  # so we are not skipping `bad_word_ids`
    continue

prefix_length = len(bad_word_ids) - 1             # 2 - 1 = 1
last_token_id = bad_word_ids[-1]                  # "world"
actual_prefix = past_tokens_ids[-prefix_length:]  # ["bye"]
expected_prefix = bad_word_ids[:prefix_length]    # ["bye"]

assert len(actual_prefix) == len(expected_prefix)  # 1 == 1

is_match = tuple(actual_prefix) == tuple(expected_prefix)  #  ["bye"] ==  ["bye"]
last_token_bias[last_token_id] += (
    self._SMALLEST_LOGIT if is_match
    else self._NEUTRAL_LOGIT)  # "world" gets float("-inf") as logit
```

So, the word "world" is prohibited.

Comment (DarkLight1337) [reply vllm/logits_process.py:70]: I see, thanks for the detailed explanation.

**Review by DarkLight1337 — APPROVED** [review]
In that case, looks good to me. Sorry for the long delay!

**Comment by Alvant** [comment]
> In that case, looks good to me. Sorry for the long delay!

All's well that ends well)

I'm really glad we were able to get everything sorted out quickly today. Thank you!

**Comment by DarkLight1337** [comment]
Please merge from main branch to fix the CI failures.
